src/algebra/algebra/spectrum.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2021 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import algebra.algebra.basic
+import tactic.noncomm_ring
+/-!
+# Spectrum of an element in an algebra
+This file develops the basic theory of the spectrum of an element of an algebra.
+This theory will serve as the foundation for spectral theory in Banach algebras.
+
+## Main definitions
+
+* `resolvent a : set R`: the resolvent set of an element `a : A` where
+  `A` is an  `R`-algebra.
+* `spectrum a : set R`: the spectrum of an element `a : A` where
+  `A` is an  `R`-algebra.
+
+## Main statements
+
+* `smul_eq_smul`: units in the scalar ring commute (multiplication) with the spectrum.
+* `left_add_coset_eq`: elements of the scalar ring commute (addition) with the spectrum.
+* `units_mem_mul_iff_mem_swap_mul` and `preimage_units_mul_eq_swap_mul`: the units
+  (of `R`) in `σ (a*b)` coincide with those in `σ (b*a)`.
+
+## Notations
+
+* `σ a` : `spectrum R a` of `a : A`
+
+## TODO
+
+* Prove the *spectral mapping theorem* for the polynomial functional calculus
+-/
+
+universes u v
+
+
+section defs
+
+variables (R : Type u) {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+-- definition and basic properties
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *resolvent* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+def resolvent (a : A) : set R :=
+{ r : R | is_unit (algebra_map R A r - a) }
+
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *spectrum* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is not a unit of the
+algebra `A`.
+
+The spectrum is simply the complement of the resolvent.  -/
+def spectrum (a : A) : set R :=
+(resolvent R a)ᶜ
+
+end defs
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+-- products of scalar units and algebra units
+
+lemma is_unit.smul_iff {G : Type u} [group G] [mul_action G A]
+  [smul_comm_class G A A] [is_scalar_tower G A A] {r : G} {a : A} :
+  is_unit (r • a) ↔ is_unit a :=
+begin
+  split, swap,
+    { exact λ h, (r•h.unit).is_unit, },
+    { intro h, let a' : units A :=
+        ⟨a,
+         r•↑(h.unit)⁻¹,
+         by rw [mul_smul_comm, ←smul_mul_assoc, is_unit.mul_coe_inv],
+         by rw [smul_mul_assoc,←mul_smul_comm, is_unit.coe_inv_mul],⟩,
+      exact ⟨a',rfl⟩, },
+end
+
+lemma is_unit.smul_sub_iff_sub_inv_smul {r : units R} {a : A} :
+  is_unit (r • 1 - a) ↔ is_unit (1 - r⁻¹ • a) :=
+begin
+  have a_eq : a = r•r⁻¹•a, by simp,
```

Comment:
This is a great technique for seeing if mathlib can solve the problem, but is the type of thing you should normally clean up once you PR to mathlib.

In particular, if you do `by simp?` you'll find that we already have a lemma for this, and it's called `smul_inv_smul`